cmd/nirvana/project/init.go

```diff
@@ -0,0 +1,782 @@
+		data = bytes.TrimSpace(data)
+		td.Boilerplate = string(data)
+	}
+
+	directories := o.directories(projectName)
+	for i, dir := range directories {
+		dir = filepath.Join(pathToProject, dir)
+		_, err := os.Stat(dir)
+		if !os.IsNotExist(err) {
+			if err != nil {
+				return fmt.Errorf("can't get stat for %s: %v", dir, err)
+			}
+			return fmt.Errorf("%s already exists", dir)
+		}
+		directories[i] = dir
+	}
+	for _, dir := range directories {
+		if err := os.MkdirAll(dir, 0775); err != nil {
+			return fmt.Errorf("can't create directory %s: %v", dir, err)
+		}
+	}
+
+	files := map[string][]byte{}
+	for file, tpl := range o.templates(projectName) {
+		tmpl, err := template.New(file).Parse(tpl)
+		if err != nil {
+			return fmt.Errorf("can't create template %s: %v", file, err)
+		}
+		buf := bytes.NewBuffer(nil)
+		if err := tmpl.Execute(buf, td); err != nil {
+			return fmt.Errorf("can't execute template %s: %v", file, err)
+		}
+		if strings.HasSuffix(file, ".go") {
+			files[file], err = format.Source(buf.Bytes())
+			if err != nil {
+				return fmt.Errorf("can't format go source file %s: %v", file, err)
+			}
+		} else {
+			files[file] = buf.Bytes()
+		}
+	}
+	for file, data := range files {
+		file = filepath.Join(pathToProject, file)
+		if err := ioutil.WriteFile(file, data, 0664); err != nil {
+			return fmt.Errorf("can't write file %s: %v", file, err)
+		}
+	}
+	return nil
+}
+
+func (o *initOptions) directories(project string) []string {
+	return []string{
+		"bin",
+		fmt.Sprintf("cmd/%s", project),
+		fmt.Sprintf("build/%s", project),
+		"pkg/api/filters",
+		"pkg/api/middlewares",
+		"pkg/api/modifiers",
+		"pkg/api/v1/descriptors",
+		"pkg/api/v1/converters",
+		"pkg/api/v1/middlewares",
+		"pkg/message",
+		"pkg/version",
+		"vendor",
+	}
+}
+
+func (o *initOptions) templates(project string) map[string]string {
+	return map[string]string{
+		fmt.Sprintf("cmd/%s/main.go", project):      o.templateMain(),
+		fmt.Sprintf("build/%s/Dockerfile", project): o.templateDockerfile(),
+		"pkg/api/filters/filters.go":                o.templateFilters(),
+		"pkg/api/middlewares/middlewares.go":        o.templateMiddlewares(),
+		"pkg/api/modifiers/modifiers.go":            o.templateModifiers(),
+		"pkg/api/v1/descriptors/descriptors.go":     o.templateDescriptors(),
+		"pkg/api/v1/descriptors/message.go":         o.templateMessageAPI(),
+		"pkg/api/v1/middlewares/middlewares.go":     o.templateMiddlewares(),
+		"pkg/api/api.go":                            o.templateAPI(),
+		"pkg/message/message.go":                    o.templateMessage(),
+		"pkg/version/version.go":                    o.templateVersion(),
+		"Gopkg.toml":                                o.templateGopkg(),
+		"Makefile":                                  o.templateMakefile(),
+		"project.yaml":                              o.templateProject(),
+	}
+}
+
+func (o *initOptions) templateMain() string {
```

Comment:
It is better to split out these templates to dedicated files for better management. Also we should provide a way to custom the default templates.

Reply:
@kdada